### PR TITLE
1.21 Scrypt algorithm

### PR DESCRIPTION
**Pull request by rnicoll**

### CODE REVIEW

**Comment by rnicoll** [comment]
Opening this early so I can track what I'm up to

**Comment by rnicoll** [comment]
Addresses #1793

**Comment by andrp92** [comment]
@rnicoll FYI, there are some conflicts with src/pow.cpp

**Comment by rnicoll** [comment]
@andrp92 Seen it! I've broken a few parts of this into smaller PRs, which I need to figure out Travis timeouts on, but in theory the plan is we'll merge the smaller parts then rebase this, and I'll resolve the conflicts then.

**Comment by andrp92** [comment]
Sounds perfect 👌

File: src/chainparams.cpp [hunk 1]
```diff
@@ -76,7 +76,7 @@ class CMainParams : public CChainParams {
         consensus.CSVHeight = std::numeric_limits<int>::max(); // Disabled on Dogecoin
         consensus.SegwitHeight = std::numeric_limits<int>::max(); // Disabled on Dogecoin
         consensus.MinBIP9WarningHeight = std::numeric_limits<int>::max(); // Dogecoin does not use BIP9
-        consensus.powLimit = uint256S("0x00000fffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"); // ~uint256(0) >> 20;
+        consensus.powLimit = uint256S("00000fffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"); // ~uint256(0) >> 20;
```
Comment (rnicoll) [inline src/chainparams.cpp:79]:
```suggestion
        consensus.powLimit = uint256S("0x00000fffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"); // ~uint256(0) >> 20;
```

File: src/chainparams.cpp [hunk 2]
```diff
@@ -200,7 +193,7 @@ class CTestNetParams : public CChainParams {
...
```
Comment (rnicoll) [inline src/chainparams.cpp:196]:
```suggestion
        consensus.powLimit = uint256S("0x00000fffffffffffffffffffffffffffffffffffffffffffffffffffffffffff"); // ~uint256(0) >> 20;
```

File: src/chainparams.cpp [hunk 3]
```diff
@@ -356,7 +332,7 @@ class SigNetParams : public CChainParams {
         consensus.nRuleChangeActivationThreshold = 1900; // 95% of 2000
         consensus.nMinerConfirmationWindow = 2000; // nPowTargetTimespan / nPowTargetSpacing
         consensus.MinBIP9WarningHeight = 0;
-        consensus.powLimit = uint256S("00000fffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
+        consensus.powLimit = uint256S("0000f77ae0000000000000000000000000000000000000000000000000000000");
```
Comment (rnicoll) [inline src/chainparams.cpp:335]:
```suggestion
        consensus.powLimit = uint256S("0x0000f77ae0000000000000000000000000000000000000000000000000000000");
```

**Review by patricklodder — CHANGES_REQUESTED** [review]

File: src/Makefile.am [hunk 4]
```diff
@@ -672,8 +675,8 @@ if GLIBC_BACK_COMPAT
 endif
 
 libbitcoinconsensus_la_LDFLAGS = $(AM_LDFLAGS) -no-undefined $(RELDFLAGS)
-libbitcoinconsensus_la_LIBADD = $(LIBSECP256K1)
-libbitcoinconsensus_la_CPPFLAGS = $(AM_CPPFLAGS) -I$(builddir)/obj -I$(srcdir)/secp256k1/include -DBUILD_BITCOIN_INTERNAL
+libbitcoinconsensus_la_LIBADD = $(LIBSECP256K1) $(CRYPTO_LIBS)
+libbitcoinconsensus_la_CPPFLAGS = $(AM_CPPFLAGS) -I$(builddir)/obj -I$(srcdir)/secp256k1/include -DBUILD_BITCOIN_INTERNAL $(SSL_CFLAGS)
```
[the remark this answers is not in the document]
Comment (rnicoll) [reply src/Makefile.am]: Seems to work without, once I realised I needed all of the later commits (as in, it won't fully work until AuxPoW is in), have removed.

File: src/net_processing.cpp [hunk 5]
```diff
@@ -2469,6 +2469,8 @@ void PeerManager::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDat
         AddTimeData(pfrom.addr, nTimeOffset);
 
         // If the peer is old enough to have the old alert system, send it the final alert.
+        // XXX: Decide how to handle alert migration in Dogecoin and either make this a Dogecoin
```
[the remark this answers is not in the document]
Comment (rnicoll) [reply src/net_processing.cpp]: Done https://github.com/dogecoin/dogecoin/issues/2231

File: test/functional/rpc_dumptxoutset.py [hunk 6]
```diff
@@ -37,13 +37,14 @@ def run_test(self):
         #           This is presumed safe none of this code changed, so we're updating the value for future changes to test against.
         assert_equal(
             out['base_hash'],
-            '3e1dc20bda7548441325d5f6224c5ae3aaa7259564540db17646cf9224fc58dd')
+            'eec7e2fd4d81b77ee3122f9500723c3723ef48aa7b45cbb54fae0c6b39f13208')
 
         with open(str(expected_path), 'rb') as f:
             digest = hashlib.sha256(f.read()).hexdigest()
+            print(str(digest))
```
[the remark this answers is not in the document]
Comment (rnicoll) [reply test/functional/rpc_dumptxoutset.py]: Done

File: src/test/miner_tests.cpp [hunk 7]
```diff
@@ -51,34 +51,34 @@ constexpr static struct {
...
+    {1, 0xe09fc8f2}, {5, 0x2ad325c5}, {5, 0x466b6549}, {1, 0x10705d49},
+    {1, 0xf88478ce}, {2, 0xbfda6c4a}, {2, 0x731fe414}, {1, 0x6f1b362e},
+    {2, 0x6be709cf}, {1, 0x60553200}, {2, 0xf6a992f0}, {2, 0x1521f7f5},
+    {1, 0x8b440273}, {1, 0xe9ade0c8}, {1, 0x4d414618}, {5, 0x7b48070d},
+    {1, 0x1202ebae}, {1, 0xd23fe97e}, {1, 0x8d1d6505}, {1, 0xfafbaae3},
+    {1, 0xf200353e}, {1, 0xe77bd65e}, {1, 0x9fa32102}, {2, 0x68dfa747},
+    {0, 0x7c74d78e}, {1, 0x9b79cc6b}, {2, 0xad957cc2}, {2, 0x91acb818},
+    {1, 0x00024b92}, {1, 0x0002a868}, {1, 0x0000767a}, {1, 0x0003f818},
+    {1, 0x001e2f24}, {1, 0x001d43f5}, {1, 0x00074756}, {2, 0x0001cc72},
+    {0, 0x0002bb2c}, {1, 0x001a9616},
```
[the remark this answers is not in the document]
Comment (rnicoll) [reply src/test/miner_tests.cpp:81]: Bitcoin Core now asks for 110 blocks ( https://github.com/bitcoin/bitcoin/blob/master/src/test/miner_tests.cpp#L223 ) so I've added these to match.

**Review by patricklodder — APPROVED** [review]
ACK. Code reviewed, syncs until 158000 for me on testnet.

**Pull request merged**